feat: add image parsing library to CEL

vishal-chdhry opened the pull request:
<!--  Thanks for sending a pull request!  Here are some tips for you:

1. If this is your first time, please read our contributor guidelines: https://git.k8s.io/community/contributors/guide/first-contribution.md#your-first-contribution and developer guide https://git.k8s.io/community/contributors/devel/development.md#development-guide
2. Please label this pull request according to what type of issue you are addressing, especially if this is a release targeted pull request. For reference on required PR/issue labels, read here:
https://git.k8s.io/community/contributors/devel/sig-release/release.md#issuepr-kind-label
3. Ensure you have added or ran the appropriate tests for your PR: https://git.k8s.io/community/contributors/devel/sig-testing/testing.md
4. If you want *faster* PR reviews, read how: https://git.k8s.io/community/contributors/guide/pull-requests.md#best-practices-for-faster-reviews
5. If the PR is unfinished, see how to mark it: https://git.k8s.io/community/contributors/guide/pull-requests.md#marking-unfinished-pull-requests
-->

#### What type of PR is this?
/kind feature
<!--
Add one of the following kinds:
/kind bug
/kind cleanup
/kind documentation


Optionally add one or more of the following kinds if applicable:
/kind api-change
/kind deprecation
/kind failing-test
/kind flake
/kind regression
-->

#### What this PR does / why we need it:

Add functions for parsing image references in CEL library

#### Which issue(s) this PR fixes:
<!--
*Automatically closes linked issue when PR is merged.
Usage: `Fixes #<issue number>`, or `Fixes (paste link of issue)`.
_If PR is about `failing-tests or flakes`, please post the related issues/tests in a comment and do not use `Fixes`_*
-->
Related #124490

#### Special notes for your reviewer:

- This CEL library does not make any network calls to registry as that would take too long to be on API server. It does not fetch information from registry such as digest, (can be useful in MAPs to mutate image digest) because that adds latency and auth related complexity.
- ~~It uses go-containerregistry's name package (https://github.com/google/go-containerregistry ) for parsing. I picked this package because:~~
	1. ~~It is a small package~~
	2. ~~It is well maintained, it is used in the [sigstore](https://github.com/sigstore) project~~
	3. ~~It is also the most popular one based on github stars~~
	4. ~~Parsing image references locally can get complicated because of old formats, backwards compatibility, locally hosted registries, proxies etc~~
- go-containerregistry's name package has multiple dependencies that are marked as unwanted in `hack/unwanted-dependencies.json`  I have replaced it with `github.com/distribution/reference` which is already in use in the project. I have also removed the default registry behaviour, now `image("nginx").registry()` returns `""`

#### Does this PR introduce a user-facing change?
<!--
If no, just write "NONE" in the release-note block below.
If yes, a release note is required:
Enter your extended release note in the block below. If the PR requires additional action from users switching to the new release, include the string "action required".

For more information on release notes see: https://git.k8s.io/community/contributors/guide/release-notes.md
-->
```release-note
New CEL Library functions for parsing image references.
```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

<!--
This section can be blank if this pull request does not require a release note.

When adding links which point to resources within git repositories, like
KEPs or supporting documentation, please reference a specific commit and avoid
linking directly to the master branch. This ensures that links reference a
specific point in time, rather than a document that may change over time.

See here for guidance on getting permanent links to files: https://help.github.com/en/articles/getting-permanent-links-to-files

Please use the following format for linking documentation:
- [KEP]: <link>
- [Usage]: <link>
- [Other doc]: <link>
-->
```docs
NONE
```

/sig api-machinery
/cc @jpbetz @cici37

cici37 commented:
Thanks for the PR. To add a new library, please also define the cost in https://github.com/kubernetes/kubernetes/blob/master/staging/src/k8s.io/apiserver/pkg/cel/library/cost.go and add tests in https://github.com/kubernetes/kubernetes/blob/master/staging/src/k8s.io/apiserver/pkg/cel/library/cost_test.go and https://github.com/kubernetes/kubernetes/blob/master/staging/src/k8s.io/apiserver/pkg/cel/library/library_compatibility_test.go
Also, enable the library in https://github.com/kubernetes/kubernetes/blob/18e5a4d585f6e3034b10b6b33f5aef436b5c49ce/staging/src/k8s.io/apiserver/pkg/cel/environment/base.go#L62 with the version info to enable the usage for all features using cel in Kubernetes.

Thanks^^

vishal-chdhry commented:
@cici37 Thank you so much for looking into this, I added all the mentioned tests.

Regarding tests in [apiextensions-apiserver/pkg/apiserver/schema/cel/celcoststability_test.go](https://github.com/kubernetes/kubernetes/blame/master/staging/src/k8s.io/apiextensions-apiserver/pkg/apiserver/schema/cel/celcoststability_test.go) , 
I have added the tests, but the test [`TestCelEstimatedCostStability`](https://github.com/vishal-chdhry/kubernetes/blob/4cbfe27305ab25a0f50a52b7092255c44bfd5d8b/staging/src/k8s.io/apiextensions-apiserver/pkg/apiserver/schema/cel/celcoststability_test.go#L1254) fails to compile on my machine. This is because `environment.DefaultCompatibilityVersion()` is set to k8s version 1.32, and, for image library, I have set `IntroducedVersion` to 1.33 in `base.go`. If I set the base version to v1.32, the test passes. 

Any suggestions on how to resolve this?

cici37 commented:
@vishal-chdhry Thanks for adding all those tests!
The failure in [apiextensions-apiserver/pkg/apiserver/schema/cel/celcoststability_test.go](https://github.com/kubernetes/kubernetes/blame/master/staging/src/k8s.io/apiextensions-apiserver/pkg/apiserver/schema/cel/celcoststability_test.go) is expected due to the way we introduce the new cel library. The new library will be able to be used in the next release(1.34) due to backward compatibility concern. It would be great if you could add the similar tests there in next release when you have time^^

The changes lgtm overall. Thank you!

/lgtm

cici37 commented:
/ok-to-test

cici37 commented:
/hold
for CI job to pass.

For the vendor updates, please run ./hack/update-vendor.sh for the new added dir and dependency

vishal-chdhry commented:
/test pull-kubernetes-e2e-gce

vishal-chdhry commented:
@cici37 

I have updated the implementation. I had to remove gcr's name package because it has multiple dependencies that are marked as unwanted in `hack/unwanted-dependencies.json` I have replaced it with `github.com/distribution/reference` which is already in use in the project. I have also removed the default registry behaviour, now `image("nginx").registry()` returns `""`

cici37 commented:
Thanks for the work!
My 2 cents: I would prefer having an option provided for normalization like what we did for semver.  :)

vishal-chdhry commented:
/test pull-kubernetes-e2e-kind

vishal-chdhry commented:
@cici37 I have added the normalization logic and tests. CI is passing now

Can you please review it

cici37 commented on staging/src/k8s.io/apiserver/pkg/cel/environment/base.go:
```diff
@@ -183,6 +183,13 @@ var baseOptsWithoutStrictCost = []VersionedOptions{
 	},
+	// Image
+	{
+		IntroducedVersion: version.MajorMinor(1, 34),
```
It should be `IntroducedVersion: version.MajorMinor(1, 33),`

cici37 commented on staging/src/k8s.io/apiserver/pkg/cel/library/image.go:
```diff
@@ -0,0 +1,312 @@
+//	image('registry.k8s.io//kube-apiserver-arm64') // error
+//	image(0) // error
+///	image('ubuntu', true) // Applies normalization and returns components of "docker.io/library/ubuntu".
+///	image('ubuntu', false) // Returns Image for "ubuntu".
```
worth mentioning that image('ubuntu') defaults to no normalization behavior which returns Image?

cici37 commented on staging/src/k8s.io/apiextensions-apiserver/pkg/apiserver/schema/cel/validation_test.go:
```diff
@@ -2147,6 +2147,19 @@ func TestValidationExpressions(t *testing.T) {
+				`image("reg.io/repo/img:tag").tag() == "tag"`,
+				`image("reg.io/repo/img@sha256:6aefddb645ee6963afd681b1845c661d0ea4c3b20ab9db86d9e753b203d385f2").digest() == "sha256:6aefddb645ee6963afd681b1845c661d0ea4c3b20ab9db86d9e753b203d385f2"`,
+			},
+		},
```
I actually don't expect those test to pass hence the library should be available for serving in the next release of introduced version. This concerns me especially it current passes with introduced version set to 1.34. @siyuanfoundation Do you happen to know if any recent emulation version related change would affect the CEL compatible version?

siyuanfoundation replied:
I am not aware of any changes that could affect CEL compat version. 
I have tried to replace the effective version with
```
func DefaultCompatibilityVersion() *version.Version {
	return version.MajorMinor(1, 33)
	// effectiveVer := compatibility.DefaultComponentGlobalsRegistry.EffectiveVersionFor(basecompatibility.DefaultKubeComponent)
	// if effectiveVer == nil {
	// 	effectiveVer = compatibility.DefaultBuildEffectiveVersion()
	// }
	// return effectiveVer.MinCompatibilityVersion()
}
```
locally, and the test still passes. So might be some other reason?

cici37 replied:
I verified that it's a issue on master. Tracked in: https://github.com/kubernetes/kubernetes/issues/130900

vishal-chdhry replied:
Should I mark this as resolved, as a fix for that issue should fix this too?

cici37 replied:
An expected fix should fail this test so the pr would need to rebase

cici37 commented on staging/src/k8s.io/apiextensions-apiserver/pkg/apiserver/schema/cel/celcoststability_test.go:
```diff
@@ -1195,6 +1195,19 @@ func TestCelCostStability(t *testing.T) {
+				`image("reg.io/repo/img@sha256:6aefddb645ee6963afd681b1845c661d0ea4c3b20ab9db86d9e753b203d385f2").digest() == "sha256:6aefddb645ee6963afd681b1845c661d0ea4c3b20ab9db86d9e753b203d385f2"`: 18,
+			},
+		},
 	}
```
Same as below. Didn't expect to pass.

jpbetz replied:
I also thought cost stability was only for fully enabled libraries.

But I attached a debugger the validator used for these tests used the stored expressions env: https://github.com/kubernetes/kubernetes/blob/ae754d050e4b6b0bab5d4a457d49cf01478f2ad1/staging/src/k8s.io/apiextensions-apiserver/pkg/apiserver/schema/cel/celcoststability_test.go#L1227

I also ran an apiserver from this PR and tried end-to-end:

```
kubectl apply -f crd.yaml
The CustomResourceDefinition "crontabs.stable.example.com" is invalid: spec.validation.openAPIV3Schema.properties[spec].properties[image].x-kubernetes-validations[0].rule: Invalid value: apiextensions.ValidationRule{Rule:"isImage(self)", Message:"", MessageExpression:"", Reason:(*apiextensions.FieldValueErrorReason)(nil), FieldPath:"", OptionalOldSelf:(*bool)(nil)}: compilation failed: ERROR: <input>:1:8: undeclared reference to 'isImage' (in container '')
 | isImage(self)
 | .......^
```

This seems OK.  It does suggest we need some additional tests that provide assurance that the bse env is working as expected.  A separate issue seems appropriate.

Last thing to consider:  Do we really need cost tests for libraries both in the library and in cost stability?  Cost stability is great for the core language and libraries we don't control.. But maybe not for libraries we control and already write separate cost testing for.

cici37 commented:
Sorry for the confusion. My bad memory on when we use stored env drives me nuts...

The changes should be ready to go. Thank you for your patience. 

/lgtm

cici37 commented:
/assign @liggitt 
if you would have time for a new cel lib together with go.mod updates :)

vishal-chdhry commented:
@cici37 @liggitt I pushed a commit to fix a comment and import hash algorithms: [`ddff560` (#130834)](https://github.com/kubernetes/kubernetes/pull/130834/commits/ddff560e4e2ebc2c110a2e25f2b1771ce310e924)

cici37 commented:
Could you squash the commits when you have time? Thanks

liggitt commented:
> Related #124490

I'm trying to get oriented here... I don't see image spec parsing mentioned in https://github.com/kubernetes/kubernetes/issues/124490 ... is there a specific KEP or feature this is associated with?

Image spec parsing is sort of a niche thing to add a CEL library for, I wasn't sure if this had been discussed elsewhere or if I was missing context.

liggitt commented:
In general, every CEL library / function we add fragments the Kubernetes ecosystem, since manifests that use that library / function will fail on older Kubernetes versions. We've typically tried to limit use of those to things which are ~impossible or extremely difficult to do with existing basic CEL validations (e.g. Kubernetes quantity parsing) or are very broadly applicable (URL / IP / CIDR parsing and intersection computation, etc).

Before we get into the mechanics of adding a library / implementing cost calculations, etc, we need to figure out whether a CEL library for this makes sense.

liggitt commented:
/hold
to figure out if this is a capability we think makes sense to add a CEL library for

liggitt commented:
cc @jpbetz for visibility / input on whether a CEL library is justified here, or if an openapi format string would be a lighter-weight more preferred way to verify a given field is a valid image spec

cici37 commented:
I could see the image library used in more area other than `verify a given field is a valid image spec`. e.g. the repository/tag/identifier/digest check could possibly used in policy enforcement area. We have similar use case for validation webhook before such like checking if the image is pulling from a specific org registry(It should have a workaround support with regex). I agree that cel library has a larger surface area than things like openapi format. Maybe a clear use case analysis would help proceeding the effort further :)

jpbetz commented:
We've been cautiously adopting support for formats where a plain OpenAPI format is insufficient for major use cases. Deconstruction (breaking apart a complex format and inspecting sub parts), and comparisons are good examples of operations beyond format validation that have motivated recent additions.

Some examples: IP and CIDR (API gateway use case in particular), semver and quantity (DRA use cases).

Because image is an important format in Kubernetes, and because it has a complex structure where deconstructing/inspecting sub parts is non-trivial.  It seems like a candidate we should seriously consider.

vishal-chdhry committed:
feat: add image parsing library to CEL

Signed-off-by: Vishal Choudhary <vishal.chdhry.work@gmail.com>

vishal-chdhry commented:
> Image spec parsing is sort of a niche thing to add a CEL library for

Hi @liggitt, I agree with joe here, Since images are integral to kubernetes, having an image parsing library makes sense,

Some of the common usecases for this library in VAPs are:
1. verifying that container images are from a specific registry/tag etc
2. ensuring that images contain digest

For MAPs
1. Replacing registries for images
2. Adding proxy registry prefixes to images from a specific repo

This library is helpful because sometimes splitting an image string is not easy using CEL.

vishal-chdhry commented:
> Since the Code Freeze is in a couple hours, trying to understand how urgent this changes are. Is there urgent need for this library to be used sooner?

Has the code freeze happened? We wanted to use this as a library in our CEL environment

jpbetz commented on staging/src/k8s.io/apiserver/pkg/cel/library/image.go:
```diff
@@ -0,0 +1,317 @@
+//
+// Image components:
+//
+//   - registry: Returns the registry component of the image reference.
+//
+//   - repository: Returns the repository component of the image reference.
+//
+//   - tag: Returns the registry component of the image reference. Defaults to latest if both tag and digest are not present.
+//
+//   - digest: Returns the digest component of the image reference.
```
@cici37 @BenTheElder Would you have a look at the components of image references that is being considered for CEL here and help think through the edge cases and what we should expose?

jpbetz commented on staging/src/k8s.io/apiserver/pkg/cel/library/image.go:
```diff
@@ -0,0 +1,317 @@
+// image("kube-apiserver-arm64:testtag").registry() // Returns ""
+// image("registry.k8s.io/kube-apiserver-arm64:latest").registry() // Returns "registry.k8s.io"
+// image("registry.k8s.io/kube-apiserver-arm64").repository() // Returns "kube-apiserver-arm64"
+// image("registry.k8s.io/kube-apiserver-arm64").tag() // Returns "latest"
```
It's interesting that this performs implicit defaulting (returning "latest" when the tag is unset).  I'm trying to get my head around the edge cases:

- Does this still return "latest" if there is also a digest set?  That could be misleading.
- Why does this perform implicit defaulting while `registry()` does not (`docker.io` is the default if the registry is not set, right?)

liggitt replied:
yeah... we don't default a missing registry to docker.io in the API layer, we plumb the image verbatim all the way down to the node, which is where the `latest` tag is added by the kubelet, and the default registry is added by the container runtime

BenTheElder replied:
TIL, I could've sworn we defaulted it to docker.io at least in kubelet before it got to CRI.

runtimes do sometimes rewrite it at the network layer too though either way

personally if I were strictly enforcing image policies about hosts etc I'd probably want that ensured in the container runtime. But a policy like "must be valid" or "must use a digest" seems reasonable to desire out at pod / daemonset / ...

liggitt replied:
> TIL, I could've sworn we defaulted it to docker.io at least in kubelet before it got to CRI.

not intentionally... it regressed a couple times and got clarified that image name should be passed down to CRI as-is

https://github.com/kubernetes/kubernetes/issues/52110#issuecomment-361214391

https://github.com/kubernetes/kubernetes/pull/58955

liggitt replied:
Any library we add should not bake in registry or tag defaulting that actually depends on the container runtime configuration. Making it visible that there is no registry or tag or digest on an image (with the possibility to return "") could be ok, so someone could write a policy that requires it be set, etc

BenTheElder replied:
Good to know, there's also an implicit default of `ubuntu` => `docker.io/library/ubuntu:latest`, I haven't seen `/library` mentioned in this thread yet.

vishal-chdhry replied:
I missed that case in the documentation. If image is not normalized (default behaviour), we pass the image as it is
```
		{
			name: "parse_non_normalized",
			expr: `image("kube-apiserver-arm64", false)`,
			expectValue: apiservercel.Image{
				ImageReference: apiservercel.ImageReference{
					Image:      "kube-apiserver-arm64",
					Registry:   "",
					Repository: "kube-apiserver-arm64",
					Tag:        "",
					Identifier: "",
				},
			},
		},
```

jpbetz commented on staging/src/k8s.io/apiserver/pkg/cel/library/image.go:
```diff
@@ -0,0 +1,317 @@
+//
+// Image components:
+//
+//   - registry: Returns the registry component of the image reference.
```
Is registry port supported?  Examples of that would be good to see

BenTheElder replied:
Runtimes support it, and IIRC this technically makes parsing ambiguous https://kind.sigs.k8s.io/docs/user/local-registry/

jpbetz commented:
> In general I'm concerned about this because image references do not have a spec we can point to, and there are some potentially surprising implicit defaulting and behaviors contained within them which make attempting to validate image references a pretty big footgun.
> 
> there isn't even a standard way to refer to the components of an image reference because there is no spec, only mimicking docker ...

So what's our guidance?  One one extreme, we could say "we don't support the image format at all".  This is a bit surprising because Kubernetes is a container orchestration system with fields in the API that accept image references.  We do publish https://kubernetes.io/docs/concepts/containers/images/#image-names after all.

Maybe one step more accepting would be "we offer a JSON Schema format type for image so CRDs can validate that a string field is a image reference".  And step farther than that might be "and we also offer a CEL library that allows you to check if a string is an image reference", but we don't allow extraction of components.

This PR takes it quite far. Note the implicit defaulting of `tag`, and the inclusion of a `identifier()` accessor.

BenTheElder commented:
> So what's our guidance? One one extreme, we could say "we don't support the image format at all". This would be a bit surprising because Kubernetes is a container orchestration system with fields in the API that accept image references. We do publish

I think the complexity and footguns already exist for users regardless but if we introduce official upstream validating APIs we need to be careful to get it right and think through the edge cases like port versus tag, defaulting, tag versus digest, etc and how we expose those.

I think we should support it, but we should get thorough review and we might need to consider having any related kubernetes.io docs revamped first, the existing docs are a bit light on these quirks. I also think we should align how we describe components of the image to our own docs and eg docker's docs to the extent possible.

https://kubernetes.io/docs/concepts/containers/images/#image-names

Seems like a good place to cover it, that currently doesn't really attempt to break down the components.

BenTheElder commented on staging/src/k8s.io/apiserver/pkg/cel/image.go:
```diff
@@ -0,0 +1,84 @@
+)
+
+type ImageReference struct {
+	Image      string `json:"image,omitempty"`
```
we should probably document these?

jpbetz replied:
Also, I'm curious why this is a JSON serializable type.  What's the use case for this?  I was expecting `.String()` on `ImageReference` to the the way this is typically serialized..

BenTheElder commented on staging/src/k8s.io/apiserver/pkg/cel/library/cost_test.go:
```diff
@@ -1191,6 +1191,70 @@ func TestSemverCost(t *testing.T) {
+		},
+		{
+			name:                "containsDigest",
+			expr:                `image("reg.io/repo/img@sha256:6aefddb645ee6963afd681b1845c661d0ea4c3b20ab9db86d9e753b203d385f2").containsDigest()`,
```
Probably at least one of these should cover `:tag@digest`

We should also probably cover the non-sha256 digests.

vishal-chdhry replied:
I have added those in [staging/src/k8s.io/apiserver/pkg/cel/library/image_test.go](https://github.com/kubernetes/kubernetes/pull/130834/files#diff-bb40882d35c8fe09c47256f825df4923942951f7d188249dc4e3aa2e360cead6R216-R220). 
I didn't add them to cost test, should I add it there too?

BenTheElder commented on staging/src/k8s.io/apiserver/pkg/cel/library/image.go:
```diff
@@ -0,0 +1,317 @@
+	//  Import the crypto sha256 algorithm for the image parser to work
+	_ "crypto/sha256"
+	//  Import the crypto/sha512 algorithm for the image parser to work with 384 and 512 sha hashes
+	_ "crypto/sha512"
```
How does this parser handle an unregistered algorithm?

https://github.com/opencontainers/distribution-spec/blob/main/spec.md#appendix
=>
https://github.com/opencontainers/image-spec/blob/v1.0.1/descriptor.md#digests

> Implementations SHOULD allow digests with unrecognized algorithms to pass validation if they comply with the above grammar. While sha256 will only use hex encoded digests, separators in algorithm and alphanumerics in encoded are included to allow for extensions. As an example, we can parameterize the encoding and algorithm as multihash+base58:QmRZxt2b1FVZPNqd8hsiykDL3TdBDeTSPX9Kv46HmX4Gx8, which would be considered valid but unregistered by this specification.

vishal-chdhry replied:
> How does this parser handle an unregistered algorithm?

The parser does not need these for converting the digest to string, but the [go-digest](https://github.com/opencontainers/go-digest#usage) package mentions that its needed for digest verification, if we choose to verify the digest in the future. Currently we are just splitting the string. We do the same in k8s [parser utils](https://github.com/kubernetes/kubernetes/blob/master/pkg/util/parsers/parsers.go).

> Implementations SHOULD allow digests with unrecognized algorithms to pass validation if they comply with the above grammar.

The parser will handle unrecognized algorithms if it is valid as per this [grammar](https://github.com/distribution/reference/blob/8c942b0459dfdcc5b6685581dd0a5a470f615bff/reference.go#L20-L24). I should add tests to ensure this is working as intended

vishal-chdhry commented:
> In general I'm concerned about this because image references do not have a spec we can point to, and there are some potentially surprising implicit defaulting and behaviors contained within them which make attempting to validate image references a pretty big footgun.

@BenTheElder I think this is the most well defined grammar we have at the moment: https://github.com/distribution/reference/blob/8c942b0459dfdcc5b6685581dd0a5a470f615bff/reference.go#L1-L26

Not sure if it is based in some standard, but we use this grammar in kubernetes for image parsing at the moment: https://github.com/kubernetes/kubernetes/blob/master/pkg/util/parsers/parsers.go

I agree that we will need a standard for image reference

rexagod commented:
/remove-sig instrumentation

BenTheElder commented:
I think the problem with only referencing a grammar is that most users won't be able to effectively go read this grammar inside the docker source code so when we try to talk about image components we don't have a reference.

we need one anyhow but personally I'd really like to see something we can point users to before handing them a way to reject pods based on it

there are a lot of footguns in misunderstanding image references (implicit latest, implicit default host, implicit default repo, tag ignored with tag@digest, lack of common understanding of how to describe the components of a reference ...)

some of those we can mitigate with careful api but overall we need a way to refer to them

I was hoping we'd get an official OCI spec for the common terminology part since that's well beyond this PR and not only relevant to Kubernetes (versus say attempting to write it up in our own docs)

vishal-chdhry commented:
Yes, having a official spec would be great, there was [some activity](https://github.com/opencontainers/tob/pull/114) around it but has since gone quite

> I'd really like to see something we can point users to before handing them a way to reject pods based on it

IIUC, we already reject pod based on the same grammar [here](https://github.com/kubernetes/kubernetes/blob/79cc5ddaa66434057544b1e518aace864c155e62/pkg/util/parsers/parsers.go#L31). If the parser fails the image wont get accepted. So the behavior of the CEL library won't be different from kubelet.
